### DISCUSSION TITLE
Custom LayoutManager

### DISCUSSION THREAD

**Discussion by StringKe** [fabricjs/fabric.js#9558] [Show and tell]
I want to use LayoutManager to realize the effect in the video.

But I don't know which methods or events I should implement in LayoutManager?

https://share.cleanshot.com/rL7RMK3X

https://github.com/fabricjs/fabric.js/assets/31089228/edbf545c-ce21-49df-bb99-258b1e7d1939

**Comment by StringKe** [comment]
@ShaMan123 Should I inherit from LayoutManager and override the layoutObject method?

**Reply by ShaMan123** [reply]
Looks like the layout in your case is derived from the vector between the selected object and the group's center
If it is only 2 objects is it simpler because you can use the vector between the 2 objects instead of the center, which might be confusing, but looks like it can be done with more objects using the center point.
You need to subclass LayoutStrategy and expose calcLayoutResult.
Then you will need to override layoutObject or layoutObjects, whatever suits you.

**Reply by ShaMan123** [reply]
There is a [readme](https://github.com/fabricjs/fabric.js/blob/master/src/LayoutManager/README.md) and comments in the code

**Reply by StringKe** [reply]
I checked the return type of calcLayoutResult, in my case I should need to return the positional relationship of two elements?

In my understanding, because there are two objects or more than one, I need to attach two different vectors for object A in the forward direction and object B in the reverse direction. How should I distinguish whether I need to attach forward or reverse vectors?

```ts
export type LayoutStrategyResult = {
  /**
   * new center point as measured by the **containing** plane (same as `left` with `originX` set to `center`)
   */
  center: Point;

  /**
   * correction vector to translate objects by, measured in the same plane as `center`
   *
   * Since objects are measured relative to the group's center, once the group's size changes we must apply a correction to
   * the objects' positions so that they relate to the new center.
   * In other words, this value makes it possible to layout objects relative to the tl corner, for instance, but not only.
   */
  correction?: Point;

  /**
   * correction vector to translate objects by as measured by the plane
   */
  relativeCorrection?: Point;

  /**
   * new width and height of the layout target
   */
  size: Point;
};
```

**Reply by ShaMan123** [reply]
I think you should first reposition the objects and then the `LayoutStrategyResult` will handle the group's size and positioning

**Reply by StringKe** [reply]
I checked all the data in detail maybe I should override the `performLayout` method to set the new position before dispatching the original logic?

**Reply by StringKe** [reply]
I tried this and in my understanding it should work fine, can you help me see what error it is getting?

```ts

import { FitContentLayout, LayoutManager, Point } from 'fabric';
import { get } from 'lodash-es';

import type { FabricObject, LayoutContext } from 'fabric';

export class TestLayout extends LayoutManager {
    relativePoint: Point = new Point(150, 150);

    constructor() {
        super(new FitContentLayout());
    }

    performLayout(context: LayoutContext) {
        const { target: group } = context;
        const { target } = get(context, 'e', {}) as { target?: FabricObject };

        if (target) {
            const targetId = target.get('id');

            if (targetId === 'AAA') {
                const bbb = group.getObjects().find((obj) => obj.get('id') === 'BBB');
                if (bbb) {
                    const deltaX = target.left - this.relativePoint.x;
                    const deltaY = target.top - this.relativePoint.y;
                    bbb.set({ left: bbb.left - deltaX, top: bbb.top - deltaY });
                }
            } else if (targetId === 'BBB') {
                const aaa = group.getObjects().find((obj) => obj.get('id') === 'AAA');
                if (aaa) {
                    const deltaX = target.left - this.relativePoint.x;
                    const deltaY = target.top - this.relativePoint.y;
                    aaa.set({ left: aaa.left + deltaX, top: aaa.top + deltaY });
                }
            }
        }

        super.performLayout(context);
    }
}

```

**Reply by ShaMan123** [reply]
create a live repro
I am guessing that `relativePoint` is incorrect
Maybe
`sendPointToPlane(new Point(150,150), undefined, group.calcTransformMatrix())`

**Reply by ShaMan123** [reply]
Note that in your implementation 150 is measured from the group's center point

**Reply by StringKe** [reply]
This fulfills my needs, I'm not sure if there will be a problem, I'll create a live repo in a bit.

If you have a better way, please let me know, thank you very much.

```
import { FitContentLayout, LayoutManager, Point } from 'fabric';
import { get } from 'lodash-es';

import type { FabricObject, LayoutContext } from 'fabric';

export class TestLayout extends LayoutManager {
    // absolute canvas point, because group size will change
    relativePoint: Point = new Point(400, 400);

    constructor() {
        super(new FitContentLayout());
    }

    performLayout(context: LayoutContext) {
        const { target: group } = context;
        const { target } = get(context, 'e', {}) as { target?: FabricObject };

        if (target) {
            const targetId = target.get('id');

            if (targetId === 'AAA') {
                const bbb = group.getObjects().find((obj) => obj.get('id') === 'BBB');
                if (bbb) {
                    const point = target.getCenterPoint();
                    const deltaX = -(point.x + target.width / 2 - this.relativePoint.x);
                    const deltaY = -(point.y + target.height / 2 - this.relativePoint.y);

                    bbb.set({ left: deltaX, top: deltaY });
                }
            } else if (targetId === 'BBB') {
                const aaa = group.getObjects().find((obj) => obj.get('id') === 'AAA');
                if (aaa) {
                    const point = target.getCenterPoint();
                    const deltaX = -(point.x + target.width / 2 - this.relativePoint.x);
                    const deltaY = -(point.y + target.height / 2 - this.relativePoint.y);

                    aaa.set({ left: deltaX, top: deltaY });
                }
            }
        }

        super.performLayout(context);
    }
}
```

**Reply by StringKe** [reply]
https://stackblitz.com/edit/web-platform-h73ybl?file=index.html

**Reply by ShaMan123** [reply]
Very cool

**Reply by StringKe** [reply]
There is a small problem, I have modified the live repo for your convenience.

Why isn't the position corrected after scheduling the triggerLayout?

I noticed some changes in the position, but not exactly correct.

**Reply by ShaMan123** [reply]
I think this issue is that once you change the position of the other object according to the active one the group center changes but that is not reflected anywhere so the layout manager uses the old center to reposition the entire group wrongly.
This is exactly the use of `correction`. Because the center of the group changes on the fly you need to figure out how to correct the position of the objects accordingly.
I am not sure...
https://stackblitz.com/edit/web-platform-b3u4ja?file=index.html

**Reply by StringKe** [reply]
Is there any way I can do an absolute position calculation?

Your modified live repo is calculated relative to group center point.

I tried to convert it to absolute position with relativePoint calculation and then again to relative position.

I'm not sure if this is possible, it doesn't work in my test scenario, maybe I'm using the wrong method.

**Reply by ShaMan123** [reply]
https://stackblitz.com/edit/web-platform-9j8cu7?file=index.html

**Reply by StringKe** [reply]
Thank you very much, for the first time I learned about the `sendPointToPlane` method.

**Comment by StringKe** [comment]
On another note, how do I make the LayoutManager act on the Canvas?

**Reply by ShaMan123** [reply]
I am not sure I understand
You want it to layout all your objects?
I guess you could do that by grouping all of them or hack the layout manager to use the canvas as the `target`, if it is useful maybe we could support it

**Reply by StringKe** [reply]
No I want to only handle elements in canvas, for example elements inside a group in canvas should not be handled, they should be managed by the group.

I want canvas to have LayoutManager as well.
Or LayoutManager should be supported as long as the elements are containers as such.

**Reply by ShaMan123** [reply]
Please describe it a bit more
The layout manager was designed to be used by Group but I don't see why it can't be used with Canvas with some adjustments

**Reply by ShaMan123** [reply]
Question is if it will be as useful

**Reply by StringKe** [reply]
> Please describe it a bit more The layout manager was designed to be used by Group but I don't see why it can't be used with Canvas with some adjustments

I think Canvas should also support LayoutManager.

The implementation of nested elements should be left to its own parent (usually Group).

**Reply by StringKe** [reply]
> Question is if it will be as useful

Very useful in specific scenarios like templates (reports, posters, product thumbnails).

Or maybe someone wants to write an html renderer using fabricjs (to handle layouts like flex grid).

**Reply by ShaMan123** [reply]
Cool, I think we should open a dedicated discussion and outline the pros etc. and what needs to change in the layout manager API (I can help)

**Reply by StringKe** [reply]
Agreed, you can create a new one where we discuss it.

**Reply by ShaMan123** [reply]
You should open it, it is your motivation and initiative
